File: minto/5주차_DP/9095.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

let cnt = 0;
let result = [];

for (let i = 1; i < input.length; i++) {
  const selectedValue = +input[i];

  function DP(num) {
    if (num === selectedValue) {
      cnt += 1;
      return;
    }
    for (let h = 1; h <= 3; h++) {
      let plusNum = num + h;
      if (plusNum <= selectedValue) {
        DP(plusNum);
      }
    }
  }

  for (let j = 1; j <= 3; j++) {
    DP(j);
  }

  result.push(cnt);
  cnt = 0;
}

console.log(result.join('\n'));

/* 
전형적인 DP 문제이다. 
각 1, 2, 3 값들의 선택에 순서를 기억하고 있기 때문에 재귀를 생각했다. 
*/

File: ujin/5주차_DP/9095.py
dp = [0] * 11
dp[1] = 1  # 1
dp[2] = 2  # 1+1 / 2
dp[3] = 4  # 1+1+1 / 1+2 / 2+1 / 3

for i in range(4, 11):
    dp[i] = dp[i - 1] + dp[i - 2] + dp[i - 3]

T = int(input())
for _ in range(T):
    n = int(input())
    print(dp[n])

'''
규칙을 이용해서 풀었다.
만약 n이 4라면
3일때의 경우들에 +1한것, 2일때의 경우들에 +2한것, 1일때의 경우들에 +3한것을 모두 합하면 된다.
'''